File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.Queue;
import java.util.LinkedList;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int time = 0;
        Queue<Integer> queue = new LinkedList<>();
        for(int tw : truck_weights) queue.add(tw);
        Queue<Integer> bridge = new LinkedList<>();
        for(int i = 0; i < bridge_length; i++) bridge.add(0);
        
        int currentWeight = 0;
        while (!queue.isEmpty()) {
            time++;
            currentWeight -= bridge.remove();
            
            if (currentWeight + queue.peek() <= weight) {
                currentWeight += queue.peek();
                bridge.add(queue.remove());
            } else {
                bridge.add(0);
            }
        }
        
        return time + bridge_length;
    }
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    time = 0
    queue = deque(truck_weights)
    bridge = deque([0] * bridge_length)
    
    currentWeight = 0
    while queue:
        time += 1
        currentWeight -= bridge.popleft()
        if currentWeight + queue[0] <= weight:
            currentWeight += queue[0]
            bridge.append(queue.popleft())
        else:
            bridge.append(0)
    
    return time + bridge_length